### PR TITLE
Fixed #22640 -- Raised TypeError when instantiating model with keyword and positional args for the same field.

### PR DESCRIPTION
**Pull request by jacobtylerwalls**
ticket-22640

### CODE REVIEW

File: django/db/models/base.py [hunk 1]
```diff
@@ -442,7 +442,10 @@ def __init__(self, *args, **kwargs):
                 if val is _DEFERRED:
                     continue
                 _setattr(self, field.attname, val)
-                kwargs.pop(field.name, None)
+                ignored_kwarg = kwargs.pop(field.name, None)
+                if ignored_kwarg:
+                    raise TypeError(f'{cls.__name__}() got both positional and '
+                                    f"keyword arguments for field '{field.name}'")
```
Comment (pope1ni) [inline django/db/models/base.py:448]: A few observations:
- There is no need for the temporary variable.
- We cannot just check if falsy as the value could be `0` or `False`.
- Compare to a sentinel object to ensure that `None` isn't passed. (I think we can re-use `NOT_PROVIDED` here.)
- Add a period to the end of the exception message.

```suggestion
                if kwargs.pop(field.name, NOT_PROVIDED) is not NOT_PROVIDED:
                    raise TypeError(
                        f'{cls.__name__}() got both positional and keyword '
                        f"arguments for field '{field.name}'."
                    )
```

**Review by pope1ni — CHANGES_REQUESTED** [review]

File: django/db/models/base.py (hunk 1 above)
Comment (jacobtylerwalls) [reply django/db/models/base.py:448]: Ouch, good catch on 0/False. Thanks for giving this a look.

Comment (jacobtylerwalls) [reply django/db/models/base.py]: Is it worth adding a second test with a Falsy value, e.g. `headline=''`?

Comment (pope1ni) [reply django/db/models/base.py]: Probably not a bad idea. You could also add, e.g. `views = IntegerField()`, to the `Article` model to test `0`, and other fields to test `False` and `None`

**Comment by felixxm** [comment]
@jacobtylerwalls Thanks for this patch :+1: , however IMO ticket-22640 hasn't been accepted to add a more descriptive error message (`__init__() got multiple values for argument 'headline'` is clear enough.), but rather to merge `*args` and `**kwargs` somehow. Preferring `*args` or `**kwargs` would be misleading and it's worth complexity, so I'm going to close this ticket.

**Comment by felixxm** [comment]
> @jacobtylerwalls Thanks for this patch +1 , however IMO [ticket-22640](https://code.djangoproject.com/ticket/22640) hasn't been accepted to add a more descriptive error message (`__init__() got multiple values for argument 'headline'` is clear enough.), but rather to merge `*args` and `**kwargs` somehow. Preferring `*args` or `**kwargs` would be misleading and it's worth complexity, so I'm going to close this ticket.

Sorry, I misread the ticket description :facepalm: .

File: django/db/models/base.py (hunk 1 above)
Comment (felixxm) [reply django/db/models/base.py]: I don't think it's worth to add multiple fields. `IntegerField(null=True)` should be enough.

File: django/db/models/base.py [hunk 2]
```diff
@@ -442,7 +442,11 @@ def __init__(self, *args, **kwargs):
                 if val is _DEFERRED:
                     continue
                 _setattr(self, field.attname, val)
-                kwargs.pop(field.name, None)
+                if kwargs.pop(field.name, NOT_PROVIDED) is not NOT_PROVIDED:
+                    raise TypeError(
+                        f'{cls.__name__}() got both positional and keyword '
```
Comment (felixxm) [inline django/db/models/base.py:447]:
```suggestion
                        f'{cls.__qualname__}() got both positional and keyword '
```

File: django/db/models/base.py (hunk 1 above)
Comment (pope1ni) [reply django/db/models/base.py]: Sure, so two extra fields - one to handle `0` and `None`, the other to handle `False`.

File: django/db/models/base.py (hunk 2 above)
Comment (pope1ni) [reply django/db/models/base.py:447]: Good shout. (Perhaps other similar exceptions could be updated in another PR.)

File: django/db/models/base.py (hunk 1 above)
Comment (felixxm) [reply django/db/models/base.py]: I think it's enough to test `None` and `''`.

**Comment by felixxm** [comment]
@jacobtylerwalls Thanks for updates :+1: I pushed small edits.

**Pull request merged**